src/sc/sidechainTxsCommitmentBuilder.cpp

```diff
@@ -0,0 +1,161 @@
+        }
+
+        field_t* certRootField = nullptr;
+        auto itCert = mScCerts.find(scid);
+        if (itCert != mScCerts.end() )
+        {
+            certRootField = itCert->second;
+        }
+
+        field_t* scIdField = mapCertToField(scid);
+
+        std::vector<field_t*> singleSidechainComponents = {ftRootField, btrRootField, certRootField, scIdField};
+        field_t * sidechainTreeRoot = merkleTreeRootOf(singleSidechainComponents);
+
+        vSortedScLeaves.push_back(sidechainTreeRoot);
+    }
+
+    field_t * finalTreeRoot = merkleTreeRootOf(vSortedScLeaves);
+    uint256 res = mapFieldToHash(finalTreeRoot);
+    zendoo_field_free(finalTreeRoot);
+
+    return res;
+}
+#endif
+
+field_t* SidechainTxsCommitmentBuilder::mapScTxToField(const uint256& ccoutHash, const uint256& txHash, unsigned int outPos) const
+{
+    static_assert((sizeof(uint256) + sizeof(uint256) + sizeof(unsigned int)) <= SC_FIELD_SAFE_SIZE);
+
+    std::string toHash = std::string(ccoutHash.begin(), ccoutHash.end()) +
+            std::string(txHash.begin(), txHash.end()) +
+            std::string(static_cast<char*>(static_cast<void*>(&outPos)));
+
+    unsigned char hash[SC_FIELD_SIZE] = {};
+    memcpy(hash, toHash.append(SC_FIELD_SIZE - toHash.size(), '\0').c_str(), SC_FIELD_SIZE);
+
+    //generate field element. MISSING null-check
+    field_t* field = zendoo_deserialize_field(hash);
+    return field;
+}
+
+field_t* SidechainTxsCommitmentBuilder::mapCertToField(const uint256& certHash) const
+{
+    static_assert(sizeof(uint256) <= SC_FIELD_SAFE_SIZE);
+    unsigned char hash[SC_FIELD_SIZE] = {};
+    std::string resHex = std::string(certHash.begin(), certHash.end());
+    memcpy(hash, resHex.append(SC_FIELD_SIZE - resHex.size(), '\0').c_str(), SC_FIELD_SIZE);
+
+    //generate field element. MISSING null-check
+    field_t* field = zendoo_deserialize_field(hash);
+    return field;
+}
+
+uint256 SidechainTxsCommitmentBuilder::mapFieldToHash(const field_t* pField) const
+{
+    if (pField == nullptr)
+        throw;
+
+    unsigned char field_bytes[SC_FIELD_SIZE];
+    zendoo_serialize_field(pField, field_bytes);
+
+    uint256 res = Hash(BEGIN(field_bytes), END(field_bytes));
+    return res;
+}
+
+unsigned int SidechainTxsCommitmentBuilder::treeHeightForLeaves(unsigned int numberOfLeaves) const
+{
+    assert(numberOfLeaves > 0);
+
+    return static_cast<unsigned int>(ceil(log2f(static_cast<float>(numberOfLeaves))));
+}
+
+field_t* SidechainTxsCommitmentBuilder::merkleTreeRootOf(std::vector<field_t*>& leaves) const
+{
+    //Notes: Leaves are consumed, i.e. freed and nulled;
+    //       It is guaranteed not to return nullptr.
+
+    unsigned int numberOfLeaves = leaves.size();
+    if (numberOfLeaves == 0) 
+        numberOfLeaves = 1;
+    
+    auto btrTree = ZendooGingerMerkleTree(treeHeightForLeaves(numberOfLeaves), numberOfLeaves);
+    for(field_t* & leaf: leaves)
+    {
+        if (leaf != nullptr) {
```

Comment:
1. previously we had a magic string hash in case of empty leaf. does current way is safe?
2. What if there is no leaves at all at some level (for example no FT and BTRs)? What is the behavior?

Reply:
1. For the moment we have assumed that if a leaf is nullptr, we just skip appending it. It's a design decision, it can be discussed (remember that this is still a draft, lot of things must change, like the fact that we will have trees of fixed height, while here we are instantiating trees of a height dependent on the number of leaves each time);

2. MerkleTree implementation in zendoo-mc-cryptolib (GingerLib) already handles the case in which you append no leaf, for a tree of whatever height. The value it returns will be a root actually computed like normal assuming that all the leaves are a "zero" field element (or some field element value you set to represent an empty leaf).

Reply:
Cool! thanks for clarifications.